op_crates/fetch/lib.rs

```diff
@@ -157,58 +157,105 @@ where
+            let (tx, rx) = mpsc::channel::<std::io::Result<Vec<u8>>>(1);
+            request = request.body(Body::wrap_stream(ReceiverStream::new(rx)));
+
+            let request_body_rid =
+              state.resource_table.add(FetchRequestBodyResource {
+                body: AsyncRefCell::new(tx),
+                cancel: CancelHandle::default(),
+              });
+
+            Some(request_body_rid)
+          }
+          Some(data) => {
+            // If a body is passed, we use it, and don't return a body for streaming.
+            request = request.body(Vec::from(&*data));
+            None
+          }
+        }
+      } else {
+        None
+      };
 
-  let mut request = client.request(method, url);
+      for (key, value) in args.headers {
+        let name = HeaderName::from_bytes(key.as_bytes()).unwrap();
+        let v = HeaderValue::from_str(&value).unwrap();
+        request = request.header(name, v);
+      }
 
-  let request_body_rid = if args.has_body {
-    match data {
-      None => {
-        // If no body is passed, we return a writer for streaming the body.
-        let (tx, rx) = mpsc::channel::<std::io::Result<Vec<u8>>>(1);
-        request = request.body(Body::wrap_stream(ReceiverStream::new(rx)));
+      let fut = request.send();
 
-        let request_body_rid =
-          state.resource_table.add(FetchRequestBodyResource {
-            body: AsyncRefCell::new(tx),
-            cancel: CancelHandle::default(),
-          });
+      let request_rid = state
+        .resource_table
+        .add(FetchRequestResource(Box::pin(fut)));
 
-        Some(request_body_rid)
-      }
-      Some(data) => {
-        // If a body is passed, we use it, and don't return a body for streaming.
-        request = request.body(Vec::from(&*data));
-        None
-      }
+      (request_rid, request_body_rid)
     }
-  } else {
-    None
-  };
+    "data" => {
+      let (bytes, content_type) = parse_data_url(&url)?;
 
-  for (key, value) in args.headers {
-    let name = HeaderName::from_bytes(key.as_bytes()).unwrap();
-    let v = HeaderValue::from_str(&value).unwrap();
-    request = request.header(name, v);
-  }
+      let response = http::Response::builder()
+        .status(http::StatusCode::OK)
+        .header(http::header::CONTENT_TYPE, content_type)
+        .body(reqwest::Body::from(bytes))?;
 
-  let fut = request.send();
+      let fut = async move { Ok(Response::from(response)) };
 
-  let request_rid = state
-    .resource_table
-    .add(FetchRequestResource(Box::pin(fut)));
+      let request_rid = state
+        .resource_table
+        .add(FetchRequestResource(Box::pin(fut)));
+
+      (request_rid, None)
+    }
+    _ => return Err(type_error(format!("scheme '{}' not supported", scheme))),
+  };
 
   Ok(FetchReturn {
     request_rid,
     request_body_rid,
   })
 }
 
+fn parse_data_url(url: &Url) -> Result<(Vec<u8>, String), AnyError> {
```

Comment:
Maybe use this for https://github.com/denoland/deno/blob/main/cli/file_fetcher.rs#L153 to avoid duplicate logic

Reply:
This now uses the `data-url` crate. I have opened issue https://github.com/denoland/deno/issues/10118 to track moving file_fetcher and lsp over to the `data-url` crate.